op-node/rollup/derive/batches.go

```diff
@@ -1,106 +1,124 @@
 package derive
 
 import (
-	"errors"
-	"fmt"
-
 	"github.com/ethereum-optimism/optimism/op-node/eth"
 	"github.com/ethereum-optimism/optimism/op-node/rollup"
 	"github.com/ethereum/go-ethereum/core/types"
 	"github.com/ethereum/go-ethereum/log"
 )
 
-var DifferentEpoch = errors.New("batch is of different epoch")
+type BatchWithL1InclusionBlock struct {
+	L1InclusionBlock eth.L1BlockRef
+	Batch            *BatchData
+}
 
-func FilterBatches(log log.Logger, config *rollup.Config, epoch eth.BlockID, minL2Time uint64, maxL2Time uint64, batches []*BatchData) (out []*BatchData) {
-	uniqueTime := make(map[uint64]struct{})
-	for _, batch := range batches {
-		if err := ValidBatch(batch, config, epoch, minL2Time, maxL2Time); err != nil {
-			if err == DifferentEpoch {
-				log.Trace("ignoring batch of different epoch", "expected_epoch", epoch,
-					"epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-			} else {
-				log.Warn("filtered batch", "expected_epoch", epoch, "min", minL2Time, "max", maxL2Time,
-					"epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions), "err", err)
-			}
-			continue
-		}
-		// Check if we have already seen a batch for this L2 block
-		if _, ok := uniqueTime[batch.Timestamp]; ok {
-			log.Warn("duplicate batch", "epoch", batch.Epoch(), "timestamp", batch.Timestamp, "txs", len(batch.Transactions))
-			// block already exists, batch is duplicate (first batch persists, others are ignored)
-			continue
-		}
-		uniqueTime[batch.Timestamp] = struct{}{}
-		out = append(out, batch)
+type BatchValidity uint8
+
+const (
+	// BatchDrop indicates that the batch is invalid, and will always be in the future, unless we reorg
+	BatchDrop = iota
+	// BatchAccept indicates that the batch is valid and should be processed
+	BatchAccept
+	// BatchUndecided indicates we are lacking L1 information until we can proceed batch filtering
+	BatchUndecided
+	// BatchFuture indicates that the batch may be valid, but cannot be processed yet and should be checked again later
+	BatchFuture
+)
+
+// CheckBatch checks if the given batch can be applied on top of the given l2SafeHead, given the contextual L1 blocks the batch was included in.
+// The first entry of the l1Blocks should match the origin of the l2SafeHead. One or more consecutive l1Blocks should be provided.
+// In case of only a single L1 block, the decision whether a batch is valid may have to stay undecided.
+func CheckBatch(cfg *rollup.Config, log log.Logger, l1Blocks []eth.L1BlockRef, l2SafeHead eth.L2BlockRef, batch *BatchWithL1InclusionBlock) BatchValidity {
+	nextTimestamp := l2SafeHead.Time + cfg.BlockTime
+	if batch.Batch.Timestamp > nextTimestamp {
+		// no log, very common case, this happens when batches are included early or out of order.
+		return BatchFuture
 	}
-	return
-}
+	// add details to the log
+	log = log.New(
```

Comment:
Is this additive in its properties or does it overwrite them? In one case, the caller passes in a logger that already has details, in another case it does not. I think the caller should be responsible for adding details to the logger before passing in

Reply:
It applies the latest properties, without mutating the logger of the caller. It copies copies the context values to a new slice within `New`, and concatenates them all with the old ones. Up to the log handler technically to pick the last context values if there are duplicates.

Reply:
See updates, de-duplicated this log context data.